src/pip/_internal/utils/entrypoints.py

```diff
@@ -25,3 +41,34 @@ def _wrapper(args: Optional[List[str]] = None) -> int:
         "running pip directly.\n"
     )
     return main(args)
+
+
+def get_best_invocation_for_this_pip() -> str:
+    """Try to figure out the best way to invoke pip in the current environment."""
+    binary_directory = "Scripts" if WINDOWS else "bin"
+    binary_prefix = os.path.join(sys.prefix, binary_directory)
+
+    # Try to use pip[X[.Y]] names, if those executables for this environment are
+    # the first on PATH with that name.
+    exe_are_in_PATH = binary_prefix in os.environ.get("PATH", "").split(os.pathsep)
+    if exe_are_in_PATH:
+        for exe_name in _EXECUTABLE_NAMES:
+            if shutil.which(exe_name) == os.path.join(binary_prefix, exe_name):
+                return exe_name
```

Comment:
I assume here that we don't care about cases where the user might, for example, have a shell alias of `pip` for some particular copy of pip - which would completely wreck all of the above logic.

Reply:
Pretty much. There's no way to detect or account for these things.